### ISSUE TITLE
Should Interval(Inf, Inf) give the empty set?

### ISSUE THREAD

**Issue by dpsanders** [JuliaIntervals/IntervalArithmetic.jl#457]
I think `Interval(Inf, Inf)` should give the empty set, rather than throwing an error as it does now.

**Comment by dpsanders** [comment]
cc @Kolaru, @lbenet

**Comment by Kolaru** [comment]
I can't see a scenario where an empty interval would give wrong result, so it looks fine to me. And the standard doesn't say much about it (p. 33):

![image](https://user-images.githubusercontent.com/17861033/116402811-21438580-a82d-11eb-96b5-ced38ccaa21f.png)

**Comment by lbenet** [comment]
The following example may show that `Interval(Inf, Inf)` is tricky as a replacement for the empty set.
```julia
julia> Interval(1,2) < Interval(Inf, Inf)
true
```
I think the standard dictates that the result should be `false`, as it currently does:
```julia
julia> emptyinterval() < Interval(1,2)
false

julia> Interval(1,2) < emptyinterval()
false
```

**Comment by lucaferranti** [comment]
If I understand correctly, the idea was that `Interval(Inf, Inf)` would _return_ an empty interval and not represent it.

At the moment the `Interval` constructor (capital i) does not perform any check, so all kind of prohibited intervals are allowed

```
julia> Interval(Inf, Inf)
[∞, ∞]

julia> Interval(2, 1)
[2, 1]
```
versus
```
julia> interval(Inf, Inf)
ERROR: ArgumentError: `[Inf, Inf]` is not a valid interval. Need `a ≤ b` to construct `interval(a, b)`.
Stacktrace:
 [1] interval(a::Float64, b::Float64)
   @ IntervalArithmetic ~\.julia\packages\IntervalArithmetic\zFxqb\src\intervals\intervals.jl:111
 [2] top-level scope
   @ REPL[27]:1

julia> interval(2, 1)
ERROR: ArgumentError: `[2, 1]` is not a valid interval. Need `a ≤ b` to construct `interval(a, b)`.
Stacktrace:
 [1] interval(a::Int64, b::Int64)
   @ IntervalArithmetic ~\.julia\packages\IntervalArithmetic\zFxqb\src\intervals\intervals.jl:111
 [2] top-level scope
   @ REPL[28]:1
```

so maybe for consistency with this `interval(Inf, Inf)` (small i) insted of `Interval(Inf, Inf)` (capital i) should return the empty set? I think that `..` and `  ±` call `interval` under the hood, so those would come for free.

**Comment by lbenet** [comment]
I agree that the issue is not to represent `emptyinterval()` as `Interval(Inf, Inf)`, but that `Interval(Inf, Inf)` is equivalent to the empty set. My point simply is a scenario in which `Interval(Inf,Inf)` may give the wrong result, following @Kolaru's [comment](https://github.com/Kolaru). 

The point is that `Interval(Inf, Inf)` in a way has to become `Interval(Inf, -Inf)` (i.e., the empty set) to avoid problems as the one I reported above.

**Comment by dpsanders** [comment]
My proposal is that there is a check  when calling `interval(Inf, Inf)`, such that the interval should then be returned rather than an error thrown. I think that deals with all the issues?

**Comment by lbenet** [comment]
Do you mean that `interval(Inf, Inf)` returns `Interval(Inf,Inf)`, keeping *different* from the empty interval? That's fine with me.

**Comment by lucaferranti** [comment]
Probably I'm just misreading something, but Section 12.1.3 of the standard (page 52) says


For some operations, and some particular inputs, there might not be a valid result. At Level 1 there are
several cases when no value exists. However, **a Level 2 operation always returns a value**. When the Level 1
result does not exist, the operation returns either
– a special value indicating this event (e.g., NaN for most of the numeric functions in 12.12.8); or
– a value considered reasonable in practice. For example, mid(Entire) returns 0; **a constructor given invalid
input returns Empty**; and one of the comparisons of 12.12.9, if any input is NaI, returns false.
The standard deﬁnes the following exceptions that may be signaled, causing a handler to be invoked:
– For the interval constructors of 12.12.7, and for exactToInterval in 13.4, it is possible that at Level
2, using ﬁnite precision, the implementation cannot decide whether a Level 1 value exists or not, see
Diﬃculties in implementation in 12.12.7. If the constructor is certain that no Level 1 value exists, the excep-
tion UndefinedOperation is signaled; if it cannot decide, the exception PossiblyUndefinedOperation is
signaled.
– If intervalPart() is called with NaI as input, the exception IntvlPartOfNaI is signaled, see 12.12.11.
– If some inputs of an operation are invalid, the exception InvalidOperand might be signaled, see 14.3.

Also section 12.12.7, where more details about the constructors are given, ends with `Otherwise the call fails, and the result is Empty.` .

So I think all constructors with not valid input such as `interval(2, 1)`, `interval(Inf, 3)` , `interval(Inf, Inf)` etc. should return ´Interval(Inf, -Inf)` .

**Comment by lucaferranti** [comment]
maybe something like 
```
function interval(a::Real, b::Real)
   !is_valid_interval(a, b) && return emptyinterval()
```
(well this is maybe not type stable,  but you get the idea)

this way `interval(Inf, Inf)` would indeed become `Interval(Inf, -Inf)` as Luis pointed out it should be.

**Comment by dpsanders** [comment]
Interesting. It does seem from the standard that invalid intervals should give the empty set.

**Comment by dpsanders** [comment]
What does the Octave interval package do?

**Comment by lucaferranti** [comment]
It prints a warning and returns an empty interval if the output is supposed to be bare, and `NaI` if the output is supposed to be decorated
```
>> a = infsup(2, 1)
warning: illegal interval boundaries: infimum greater than supremum
warning: called from
    infsup at line 829 column 5

a = [Empty]
>> a
a = [Empty]
>> b = infsupdec(2, 1)
warning: illegal interval boundaries: infimum greater than supremum
warning: called from
    infsup at line 829 column 5
    infsupdec at line 244 column 40

b = [NaI]
>> b
b = [NaI]
```

**Comment by lucaferranti** [comment]
A few more examples
```
>> infsup(Inf, Inf)
warning: illegal interval boundaries: infimum = supremum = +/- infinity
warning: called from
    infsup at line 813 column 5

ans = [Empty]
>> infsup(NaN, 2)
warning: input contains NaN values
warning: called from
    infsup at line 767 column 7

ans = [Empty]
>> infsup(NaN, NaN)
warning: input contains NaN values
warning: called from
    infsup at line 767 column 7

warning: input contains NaN values
warning: called from
    infsup at line 767 column 7

ans = [Empty]
```

**Comment by dpsanders** [comment]
OK, interesting. I guess we should do the same.
